File: src/Types/Team.hs
-- | Module: Types.Team
module Types.Team where

-- | An enumerable type containing every team in the league plus legends
data Team
  = Legends
  | Bears
  | Bengals
  | Bills
  | Broncos
  | Browns
  | Buccaneers
  | Cardinals
  | Chargers
  | Chiefs
  | Colts
  | Commanders
  | Cowboys
  | Dolphins
  | Eagles
  | Falcons
  | Giants
  | Jaguars
  | Jets
  | Lions
  | Niners
  | Packers
  | Panthers
  | Patriots
  | Raiders
  | Rams
  | Ravens
  | Saints
  | Seahawks
  | Steelers
  | Texans
  | Titans
  | Vikings
  | ULOline
  deriving (Eq, Ord, Show, Read)

thresholds :: [(Team -> Bool, Int)]
thresholds = [
    ((== Legends), 40),
    ((/= Legends), 50),
    ((== ULOline), 8),
    ((/= Legends), 40)
  ]

overAThreshold ::
  Team ->
  Team ->
  Int ->
  Int ->
  Bool
overAThreshold = overAThreshold' thresholds

overAThreshold' ::
  [(Team -> Bool, Int)] ->
  Team ->
  Team ->
  Int ->
  Int ->
  Bool
overAThreshold' [] _ _ _ _ = False
overAThreshold' ((f, n):fns) t1 t2 i1 i2 =
  and [f t1, f t2, i1 >= n, i2 >= n] || overAThreshold' fns t1 t2 i1 i2

compareThresholds ::
  Team ->
  Team ->
  Int ->
  Int ->
  Either Bool Ordering
compareThresholds = compareThresholds' thresholds

compareThresholds' :: 
  [(Team -> Bool, Int)] ->
  Team ->
  Team ->
  Int ->
  Int ->
  Either Bool Ordering
compareThresholds' [] _ _ _ _ = Left False
compareThresholds' ((f, n):fns) t1 t2 i1 i2 =
  case (f t1 && i1 >= n, f t2 && i2 >= n) of
    (True, False) -> Right GT
    (False, True) -> Right LT
    _ -> compareThresholds' fns t1 t2 i1 i2

File: src/team/team.py
"""A list of teams in the NFL for chemistry purposes"""
from enum import Enum


class Team(Enum):
    """A list of teams in the NFL for chemistry purposes"""

    NO_TEAM = "No Team"
    BEARS = "Bears"
    BENGALS = "Bengals"
    BILLS = "Bills"
    BRONCOS = "Broncos"
    BROWNS = "Browns"
    BUCCANEERS = "Buccaneers"
    CARDINALS = "Cardinals"
    CHARGERS = "Chargers"
    CHIEFS = "Chiefs"
    COLTS = "Colts"
    COMMANDERS = "Commanders"
    COWBOYS = "Cowboys"
    DOLPHINS = "Dolphins"
    EAGLES = "Eagles"
    FALCONS = "Falcons"
    GIANTS = "Giants"
    JAGUARS = "Jaguars"
    JETS = "Jets"
    LIONS = "Lions"
    NINERS = "49ers"
    PACKERS = "Packers"
    PANTHERS = "Panthers"
    PATRIOTS = "Patriots"
    RAIDERS = "Raiders"
    RAMS = "Rams"
    RAVENS = "Ravens"
    SAINTS = "Saints"
    SEAHAWKS = "Seahawks"
    STEELERS = "Steelers"
    TEXANS = "Texans"
    TITANS = "Titans"
    VIKINGS = "Vikings"
    ALL32 = "All32Teams"

    @staticmethod
    def all_teams() -> list["Team"]:
        """All the actual teams"""
        return [t for t in list(Team) if t not in Team.not_real_teams()]

    @staticmethod
    def not_real_teams() -> list["Team"]:
        """All the not-real teams, so all32 and no team (for now)"""
        return [Team.ALL32, Team.NO_TEAM]
